File: logic/financial_statements.py
from models import FinancialStatements, Transaction
from datetime import date
from typing import List

def generate_financial_statements(as_of: date, txs: List[Transaction]) -> FinancialStatements:
    # Simple balance sheet: assets = cash (assume starting cash + txs), liabilities = 0, equity = retained earnings
    cash = 10000  # Assume starting cash
    for tx in txs:
        if tx.date <= as_of:
            cash += tx.amount

    retained_earnings = cash - 10000  # Simplified

    balance_sheet = {
        "assets": {"cash": round(cash, 2)},
        "liabilities": {},
        "equity": {"retained_earnings": round(retained_earnings, 2)}
    }

    # P&L: revenue - expenses
    revenue = sum(tx.amount for tx in txs if tx.amount > 0 and tx.date <= as_of)
    expenses = sum(abs(tx.amount) for tx in txs if tx.amount < 0 and tx.date <= as_of)
    net_income = revenue - expenses

    profit_loss = {
        "revenue": round(revenue, 2),
        "expenses": round(expenses, 2),
        "net_income": round(net_income, 2)
    }

    # Cash flow: simplified as net income
    cash_flow = {
        "operating": round(net_income, 2),
        "investing": 0,
        "financing": 0,
        "net_change": round(net_income, 2)
    }

    return FinancialStatements(
        as_of_date=as_of,
        balance_sheet=balance_sheet,
        profit_loss=profit_loss,
        cash_flow=cash_flow
    )

File: lib/financialStatements.js
// lib/financialStatements.js
export function generateFinancialStatements(asOf, transactions) {
  const filtered = transactions.filter(t => t.date <= asOf);
  const revenue  = filtered.filter(t => t.amount > 0).reduce((s,t) => s + t.amount, 0);
  const expenses = filtered.filter(t => t.amount < 0).reduce((s,t) => s + Math.abs(t.amount), 0);
  const net      = revenue - expenses;
  const cash     = 10000 + net;
  return {
    as_of_date: asOf,
    balance_sheet: {
      assets: { cash: r(cash) },
      liabilities: {},
      equity: { retained_earnings: r(net) },
    },
    profit_loss: { revenue: r(revenue), expenses: r(expenses), net_income: r(net) },
    cash_flow: { operating: r(net), investing: 0, financing: 0, net_change: r(net) },
  };
}

function r(n) { return Math.round(n * 100) / 100; }
